distml/operator/jax_operator.py

```diff
@@ -0,0 +1,338 @@
+        # Should be set by users in the `register` function.
+        # model methods
+        self.opt_state = None
+        self.init_fun = None
+        self.predict_fun = None
+        # optimizer methods
+        self.opt_init = None
+        self.opt_update = None
+        self.get_params = None
+
+        self.criterion = None
+
+        # Data loaders for training and validation, registered by users.
+        self._train_loader = None
+        self._validation_loader = None
+
+        self.setup(operator_config)
+
+        if hasattr(operator_config, "jit_mode"):
+            if operator_config["jit_mode"]:
+                raise NotImplementedError("Not support jit in jax operator.")
+
+        self.train_step_num = 0
+
+    def setup(self, *args, **kwargs):
+        """Function that needs to be override by users.
+
+        Example:
+            # some code is the same for all users,
+            # maybe we can put it in register.
+            rng_key = random.PRNGKey(0)
+            input_shape = (28, 28, 1, batch_size)
+            lr=0.01
+            init_fun, predict_fun = ResNet18(num_classes)
+            _, init_params = init_fun(rng_key, input_shape)
+
+            opt_init, opt_update, get_params = optimizers.adam(lr)
+            opt_state = opt_init(init_params)
+
+            criterion = lambda logits, targets:-jnp.sum(logits * targets)
+
+            self.register(model=(opt_state, init_fun, predict_fun),
+                          optimizer=(opt_init, opt_update, get_params),
+                          criterion=criterion)
+        """
+        raise NotImplementedError("Please override this function to register "
+                                  "your model, optimizer, and criterion.")
+
+    def register(self,
+                 *,
+                 model,
+                 optimizer,
+                 criterion,
+                 jit_mode=False):
+        """Register a few critical information about the model to operator."""
+        self.criterion = criterion
+        self._register_model(model)
+        self._register_optimizer(optimizer)
+
+    def _register_model(self, model):
+        """register model components.
+
+         This function shall be instantiated in framework-specific operator
+         implementations.
+         """
+        self.opt_state = model[0]
+        self.init_fun = model[1]
+        self.predict_fun = model[2]
+
+    def _register_optimizer(self, optimizer):
+        self.opt_init = optimizer[0]
+        self.opt_update = optimizer[1]
+        self.get_params = optimizer[2]
+
+    def register_data(self, *, train_loader=None, validation_loader=None):
+        self._train_loader = train_loader
+        self._validation_loader = validation_loader
+
+    def _get_train_loader(self):
+        return self._train_loader
+
+    def _get_validation_loader(self):
+        return self._validation_loader
+
+    def loss_func(self, params, batch):
```

Comment:
where are `params` for here?  please add docstring for it below. Also might need to check whether params are valid inputs.